File: RotateLinkedList/RotateLinkedList.cpp
// RotateLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (pHead == nullptr)
        return;

    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

int CountElementsAndMoveToTail(ListNode** pHead)
{
    if (*pHead == nullptr)
        return 0;

    int count = 1;
    while ((*pHead)->next != nullptr)
    {
        ++count;
        *pHead = (*pHead)->next;
    }

    return count;
}

ListNode* RotateListByK(ListNode* head, int k)
{
    if (head == nullptr)
        return head;

    ListNode* pLast = head;
    int count = CountElementsAndMoveToTail(&pLast);
    k = k % count;

    if (k != 0)
    {
        ListNode* pNode = GetNodeByIndex(head, count - k - 1);
        pLast->next = head;
        head = pNode->next;
        pNode->next = nullptr;
    }

    return head;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count < 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        std::cout << "The number of nodes to rotate: ";
        int k = -1;
        std::cin >> k;

        pHead = RotateListByK(pHead, k);

        std::cout << "The list after rotated: ";
        PrintList(pHead);
        std::cout << std::endl;

        FreeList(pHead);
    }
}

File: Go/RotateLinkedList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) (*ListNode, *ListNode) {
	if len(v) == 0 {
		return nil, nil
	}

	if len(v) == 1 {
		pNode := &ListNode{v[0], nil}
		return pNode, pNode
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead, pCur
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
	}
	if pNode.next != nil {
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func GetNodeByIndex(head *ListNode, index int) *ListNode {
	for i := 0; i < index; i++ {
		head = head.next
	}

	return head
}

func CountListMoveToTail(head *ListNode) (int, *ListNode) {
	count := 1
	for ; head.next != nil; head = head.next {
		count++
	}

	return count, head
}

func RotateListByK(head *ListNode, k int) *ListNode {
	count, pTail := CountListMoveToTail(head)
	k = k % count

	if k != 0 {
		pNode := GetNodeByIndex(head, count-k-1)
		pTail.next = head
		head = pNode.next
		pNode.next = nil
	}

	return head
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead, _ := GetListFromArray(vec)

		fmt.Print("Please enter number k (to rotate): ")
		var k int
		fmt.Scan(&k)
		pHead = RotateListByK(pHead, k)

		fmt.Print("The rotated list: ")
		pHead.Print()
	}
}
